design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,474 @@
+
+This is problematic for us at this stage because now the Secrets Provider is sitting separately and will need to perform a batch retrieval of all the secrets in the namespace. In other words, if the Secrets Provider fails to fetch one Conjur secret, the whole request will fail and applications will not get their secrets. 
+
+To avoid this, we will parse over each K8s Secret and perform a batch retrieval request. That way, if there is a failure in retrieving a Conjur secret (a failure in permissioning for example), it will be contained to that secret and not impact all the secrets in the namespace.
+
+```pseudocode
+for each k8s-secret
+  variable-ids = k8s-secret.parseConjurMap() # list of paths of key in ConjurMap
+  GET /secrets{?variable-ids}
+  # Log failure or success
+```
+
+Variable IDs will resemble the following:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-credentials
+type: Opaque
+stringData:
+  conjur-map: |-
+    username: secrets-accessors/db_username # the value is one element in the variable-ids list
+    password: secrets-accessors/db_password # the value is one element in the variable-ids list
+```
+
+#### Fetching Relevant K8s Secrets
+
+With the decoupling of Secrets Provider and application pod, we can know what types of K8s Secrets the pod requires by either of the following ways: 
+
+|      | Fetch/Update K8s Secret                                      | Pros                                                         | Cons                                        | Effort estimation |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------- | ----------------- |
+| 1    | Read all K8s Secrets that are marked with labels specified in `K8S_SECRETS_LABEL` env var | - Scalable <br /><br />- Less intervention on our part as a customer can define which types of secrets we should parse<br /><br />- K8s native, as-is solution | - Not generalized so we are still K8s bound | 3 days            |
+| 2    | Get K8s secrets with the names specified in `K8S_SECRETS`    | - No code changes<br /><br />- Backwards compatible          | - Repetitive work<br /><br />- Not scalable | Free              |
+
+*Decision*: Solution #1, Read all K8s Secrets that are marked with labels. It would be the most K8s native and avoids the repetitive work of having to list all the K8s Secrets each app needs.
+
+Listing each K8s Secret in the ENV of the Secrets Provider manifest is not a scalable solution, especially for customers with hundreds of K8s Secrets. Customers might already use labels for grouping related K8s Secrets so we will benefit from this enhancement by integrating more with a customer environment and utilizing that.
+
+Even with this new addition, for backwards compatibility, we will still be supporting customers using `K8S_SECRETS`. 
```

Comment:
how? can you describe what we will do? will we get all the secrets in `K8S_SECRETS` and also get all the secrets that have the label?

Reply:
Added
```
K8S_SECRETS exists in manifest?
  Yes? Fetch k8s secrets defined under K8S_SECRETS
  No? K8S_SECRETS_LABEL exists?
    Yes? Fetch k8s secrets with label defined under K8S_SECRETS_LABEL
    No? Fail and log
```

Reply:
what about a default value to `K8S_SECRETS_LABEL` so they don't need to add it as an env-var?
for example, `conjur-secret:true`.

Reply:
I am against default values here because what would be the difference between  `K8S_SECRETS_LABEL` and `K8S_SECRETS` then

Reply:
there is no connection between the 2. Having a default value for `K8S_SECRETS_LABEL` will enable users to define a secret like this:
```
apiVersion: v1
kind: Secret
metadata:
  name: test-k8s-secret
  labels:
    conjur-secret: true
type: Opaque
stringData:
  conjur-map: |-
    secret: secrets/test_secret
    var_with_spaces: secrets/var with spaces
    var_with_pluses: secrets/var+with+pluses
  non-conjur-key: some-value
```

and then they don't need to add any env var (from `K8S_SECRETS` & `K8S_SECRETS_LABEL`) to the manifest.

Reply:
But the benefit is two fold
1.  The customer can use labels that already exist in their environment and 
2. Logical Grouping
`conjur-secret:true` answers the logical grouping but doesnt answer the first point that a configurable label offers us

Reply:
i know but what if the customer doesn't already have a label they use? having the default label gives them the ability to add the label to the required secrets without the need to add an env var to the secrets-providers manifests

Reply:
I agree with @orenbm I think default values are great! usually they provide the need for most of the customers. K8S_SECRETS_LABEL provide them extra flexibility but will serve more complex scenarios.
BTW this is how we should document it also - using the default value and only at the end of the flow discuss the more complex flexibility options.